--- a/csrc/kernels/intranode.cu
+++ b/csrc/kernels/intranode.cu
@@ -818,2 +818,2 @@
-                    for (int j = 0; j < num_topk_ranks; ++ j)
-                        recv_value_int4[j] = ld_nc_global(channel_x_buffers[topk_ranks[j]].buffer() + slot_indices[j] * hidden_int4 + i);
+                    for (int j = 0; j < num_topk_ranks; ++ j)
+                        recv_value_int4[j] = ld_nc_global(channel_x_buffers[0].buffer() + (topk_ranks[j] * num_recv_buffer_tokens + slot_indices[j]) * hidden_int4 + i);
@@ -873,2 +873,2 @@
-                    for (int i = 0; i < num_topk_ranks; ++ i)
-                        value += ld_nc_global(channel_topk_weights_buffers[topk_ranks[i]].buffer() + slot_indices[i] * num_topk + lane_id);
+                    for (int i = 0; i < num_topk_ranks; ++ i)
+                        value += ld_nc_global(channel_topk_weights_buffers[0].buffer() + (topk_ranks[i] * num_recv_buffer_tokens + slot_indices[i]) * num_topk + lane_id);
